spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/BinderFactoryConfiguration.java

```diff
@@ -55,30 +55,34 @@
 		Map<String, BinderConfiguration> binderConfigurations = new HashMap<>();
-		if (!CollectionUtils.isEmpty(channelBindingServiceProperties.getBinders())) {
-			for (Map.Entry<String, BinderProperties> binderEntry :
-					channelBindingServiceProperties.getBinders().entrySet()) {
-				BinderProperties binderProperties = binderEntry.getValue();
-				if (binderTypeRegistry.get(binderEntry.getKey()) != null) {
-					binderConfigurations.put(binderEntry.getKey(),
-							new BinderConfiguration(binderTypeRegistry.get(binderEntry.getKey()),
-									binderProperties.getEnvironment(), binderProperties.isInheritEnvironment()));
-				}
-				else {
-					Assert.hasText(binderProperties.getType(), "No 'type' property present for custom " +
-							"binder " + binderEntry.getKey());
-					BinderType binderType = binderTypeRegistry.get(binderProperties.getType());
-					Assert.notNull(binderType, "Binder type " + binderProperties.getType() + " is not defined");
-					binderConfigurations.put(binderEntry.getKey(),
-							new BinderConfiguration(binderType, binderProperties.getEnvironment(),
-									binderProperties.isInheritEnvironment()));
-				}
+		Map<String, BinderProperties> declaredBinders = channelBindingServiceProperties.getBinders();
+		boolean hasUserDefinedBinders = false;
+		Iterator<Map.Entry<String, BinderProperties>> binderPropertiesIterator = declaredBinders.entrySet().iterator();
+		while (!hasUserDefinedBinders && binderPropertiesIterator.hasNext()) {
+			hasUserDefinedBinders = binderPropertiesIterator.next().getValue().isApplicationProvided();
+		}
+		for (Map.Entry<String, BinderProperties> binderEntry : declaredBinders.entrySet()) {
+			BinderProperties binderProperties = binderEntry.getValue();
+			if (binderTypeRegistry.get(binderEntry.getKey()) != null) {
+				binderConfigurations.put(binderEntry.getKey(),
+						new BinderConfiguration(binderTypeRegistry.get(binderEntry.getKey()),
+								binderProperties.getEnvironment(), binderProperties.isInheritEnvironment(),
+								binderProperties.isApplicationProvided()));
+			}
+			else {
+				Assert.hasText(binderProperties.getType(),
+						"No 'type' property present for custom " + "binder " + binderEntry.getKey());
+				BinderType binderType = binderTypeRegistry.get(binderProperties.getType());
+				Assert.notNull(binderType, "Binder type " + binderProperties.getType() + " is not defined");
+				binderConfigurations.put(binderEntry.getKey(),
+						new BinderConfiguration(binderType, binderProperties.getEnvironment(),
+								binderProperties.isInheritEnvironment(), binderProperties.isApplicationProvided()));
 			}
 		}
-		else {
+		if (!hasUserDefinedBinders) {
 			for (Map.Entry<String, BinderType> entry : binderTypeRegistry.getAll().entrySet()) {
 				binderConfigurations.put(entry.getKey(),
-						new BinderConfiguration(entry.getValue(), new Properties(), true));
+						new BinderConfiguration(entry.getValue(), new Properties(), true, true));
```

Comment:
Yes. I'm just struggling with `defaultCandidate` - since it's having two meanings "is the *default* candidate" or "is this a candidate for defaulting" ?